Work-in-progress NuttX.html conversion; Pipenv; and installation/contributing pages

adamfeuer opened the pull request:
## Description

- Work-in-progress NuttX.html conversion 
- Pipenv
- Installation/contributing RST pages

## Detail

- The NuttX.html structure has been converted
- Table of Contents links work

## Limitations/TODO

- Most of the other internal links don't work yet
- Formatting for the boards table doesn't work 

## Type of change

Delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Manually via sphinx make html

**Test Configuration**:

- docs only

## Checklist:

- mostly does not apply to Sphinx-only changes

adamfeuer commented:
@v01d @btashton Here's a work in progress PR. I'll keep working on the `Nuttx.html` page, it's `introduction/main.rst` in the PR, and will submit another PR when I get the tables and internal links working.

v01d commented:
I will try to build locally and have a look since for now there's now way to preview the PR.
BTW, what is the pipenv file for? Required python packages are listed in requirements.txt (at least that's what the CI system uses to build). Also, there's a lock file which should not be there.
Note there's a specific "contributing" directory where I would place everything related on how to contribute to NuttX, including documentation. If you don't have this on your clone, please rebase on latest commit of the docs branch, so that you can place it there (there are a few stub files).

v01d commented on doc/introduction/index.rst:
```diff
@@ -1,4 +1,31 @@
 ============
 
-This should have most of the context of the main NuttX HTML which lists all supported platforms and main features.
+.. toctree::
```
there is already a toctree below, this is a duplicate

adamfeuer replied:
@v01d Oops, will remove.

v01d commented on doc/introduction/index.rst:
```diff
@@ -1,4 +1,31 @@
+   resources
+   contributing
+
+Indices and tables
```
This should be removed, it introduces a section in the sidebar and also is not very useful in itself

adamfeuer replied:
@v01d Will remove. What about links to Index and Search? Where should they go? It seems like they're both useful.

v01d replied:
There's a search box on the sidebar already. Not sure about the index, it comes up empty to me here. I think the sidebar already exposes the structure to the documentation so I would not add much more than that.

v01d commented on doc/introduction/index.rst:
```diff
@@ -1,4 +1,31 @@
+These pages are meant to be an experiment in replacing the `Apache NuttX Documentation <https://cwiki.apache.org/confluence/display/NUTTX/Nuttx>`_
+with something easier to navigate, extend, and modify.
+
+This is a work in progress, many links don't work and a lot of formatting is broken.
```
this can be changed into a warning by using `.. warning::`

adamfeuer replied:
@v01d will change to `.. warning::`

v01d commented on doc/introduction/main.rst:
```diff
@@ -0,0 +1,4440 @@
+NuttX RTOS
+==========
+
+       Last Updated: April 16, 2020
```
These dates could be generated by sphinx from latest commit information so I would remove them

adamfeuer replied:
@v01d Will change to a timestamp automatically generated by Sphinx

v01d replied:
I was thinking on adding a footer to all pages or something similar. Not sure if it is prominent enough.

v01d commented on doc/introduction/main.rst:
```diff
@@ -0,0 +1,4440 @@
+.. _main:
+
+===================
+NuttX Documentation
```
the name of this section should be changed to reflect the content of the document. since this is below "introduction" it could be something like About NuttX or NuttX RTOS (and remove the below section).

adamfeuer replied:
@v01d Will fix

v01d commented on doc/introduction/main.rst:
```diff
@@ -0,0 +1,4440 @@
+Table of Contents
+-----------------
+
+* :ref:`overview`
```
I think it would be cleaner to have the toc be built manually. You can look at how I converted the "coding style" document  (under contributing/) using the ".. contents::" directive. It also generates back-references automatically.

adamfeuer replied:
@v01d I will look at this and see if I can use the same approach.

adamfeuer replied:
@v01d Ok, I looked at your TOC approach and like it. 

My proposal is to convert this file from HTML trying to keep it mostly looking the same, and once that is done, refactor it. It's just easier for me to do the two operations serially, instead of at the same time. 

Are you ok accepting the file the way it is now? I'll make another PR to get the rest of the conversion finished. Then we can refactor to use your TOC approach, which will change the TOC appearance.

v01d replied:
Sounds good!

v01d commented on doc/introduction/main.rst:
```diff
@@ -0,0 +1,4440 @@
+
+NuttX Discussion Group
+----------------------
+
```
I would remove this, it should be present elsewhere (besides this pointing to the old group)

adamfeuer replied:
@v01d Yeah, I was just translating the document as-is. I'll remove.

v01d commented on doc/introduction/main.rst:
```diff
@@ -0,0 +1,4440 @@
+
+.. _downloads:
+
+Downloads
```
This is also obsolete and maybe not appropriate for the "intro" page

adamfeuer replied:
@v01d Will remove.

v01d commented on doc/introduction/main.rst:
```diff
@@ -0,0 +1,4440 @@
+
+.. _supported-platforms:
+
+Supported Platforms
```
I would move this to a separate rst file since it is quite long already.

adamfeuer replied:
@v01d Let's leave for now. Once I get the file translated pretty-much as-is, we can refactor. That ok with you?

v01d replied:
Sure

adamfeuer commented:
@v01d The `Pipenv` and `Pipenv.lock` files are for repeatable creation of pip virtual environments, for CI and to make it easier for people to get up and running quickly. The `Pipenv.lock` file is meant to be there– it should be checked into source control to provide a frozen set of requirements that enable repeatable builds. The Pipfile allows having unfrozen requirements to let you keep pace with the Internet as new versions are released. See https://pipenv.pypa.io/en/latest/

The problem with just using a `requirements.txt` file is that it makes you either keep your requirements all frozen, keeping the project static; or unfrozen, leading to unrepeatable or broken CI runs when new package versions come out.

I'll move the Contributing stuff to the `contributing/` dir.

v01d commented:
> @v01d The `Pipenv` and `Pipenv.lock` files are for repeatable creation of pip virtual environments, for CI and to make it easier for people to get up and running quickly. The `Pipenv.lock` file is meant to be there– it should be checked into source control to provide a frozen set of requirements that enable repeatable builds. The Pipfile allows having unfrozen requirements to let you keep pace with the Internet as new versions are released. See https://pipenv.pypa.io/en/latest/
> 
> The problem with just using a `requirements.txt` file is that it makes you either keep your requirements all frozen, keeping the project static; or unfrozen, leading to unrepeatable or broken CI runs when new package versions come out.

Oh, ok. My python knowledge is very basic. I guess CI should be updated to use that instead then and requirements.txt removed.

adamfeuer commented:
@v01d Re: `pipenv` versus `requirements.txt`, no problem. I will look at the CI and make a PR that updates it.

adamfeuer commented:
@v01d Ok, I pushed changes to the branch addressing most of your suggestions. The ones I didn't do are left above with questions. Since this is a work-in-progress PR to help people get an idea of what things will look like, my proposal is to merge as-is, and I'll do follow-up PRs to finish converting this file, and then to restructure it if that's wanted.

What do you think?

btashton commented:
> > @v01d The `Pipenv` and `Pipenv.lock` files are for repeatable creation of pip virtual environments, for CI and to make it easier for people to get up and running quickly. The `Pipenv.lock` file is meant to be there– it should be checked into source control to provide a frozen set of requirements that enable repeatable builds. The Pipfile allows having unfrozen requirements to let you keep pace with the Internet as new versions are released. See https://pipenv.pypa.io/en/latest/
> > 
> > The problem with just using a `requirements.txt` file is that it makes you either keep your requirements all frozen, keeping the project static; or unfrozen, leading to unrepeatable or broken CI runs when new package versions come out.
> 
> Oh, ok. My python knowledge is very basic. I guess CI should be updated to use that instead then and requirements.txt removed.

No this will break the build. The action requires the requirements.txt file.  I also use pipenv, but you can just feed it the requirements file. I know this let's you do better pinning, but for this kind of project I would not be too worried about that.

v01d commented:
As per @adamfeuer suggestion, I will merge this as-is so that it can be iterated in a separate PR.

BTW: I will try to finish the (tedious and manual) conversion of the C Coding Style. I think once the work and the intro and on this is done, it would be good to present it on the list before continuing with further conversions.

adamfeuer commented:
> No this will break the build. The action requires the requirements.txt file. I also use pipenv, but you can just feed it the requirements file. I know this let's you do better pinning, but for this kind of project I would not be too worried about that.

@btashton @v01d Ok. Can we leave the pipenv files in there? It'll make it easier to develop with. We can generate the `requirements.txt` file using `pipenv lock -r > requirements.txt` when we update dependencies.
